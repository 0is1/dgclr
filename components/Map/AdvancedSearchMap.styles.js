import styled from 'styled-components';
import { Button, Box } from 'rebass';
import colors from 'components/colors';

const Styles = {
  LocationButton: styled(Button)`
    align-items: center;
    background-color: ${colors.lighterGray};
    border-radius: 10px;
    border: 1px solid ${colors.grayAlpha};
    bottom: 30px;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    height: 30px;
    justify-content: center;
    left: 10px;
    padding: 1rem;
    position: absolute;
    width: 30px;
    &:focus {
      box-shadow: 0 0 0 2px ${colors.grayAlpha};
    }
  `,
  LocationIconWrapper: styled.span`
    width: 20px;
    > svg {
      color: ${props => (props.useCurrentLocation ? colors.pink : colors.gray)};
      width: 15px;
    }
  `,
  ErrorWrapperBox: styled(Box)`
    align-items: center;
    background-color: ${colors.white};
    margin: 1rem 0;
    display: flex;
    flex-direction: row;
    justify-content: center;
    svg {
      color: ${colors.red};
      margin-right: 1rem;
    }
  `,
};

export default Styles;
