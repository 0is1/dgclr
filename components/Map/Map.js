// @flow

import React, { PureComponent } from 'react';
import {
  compose, lifecycle, withProps, withStateHandlers,
} from 'recompose';
import {
  withScriptjs, withGoogleMap, GoogleMap, Marker, Circle,
} from 'react-google-maps';
import InfoBox from 'react-google-maps/lib/components/addons/InfoBox';
import { BarLoader } from 'components/Spinners';
import colors from 'components/colors';
import type { CoordinatesObject } from 'lib/types';

const MyMapComponent = compose(
  withProps({
    googleMapURL: 'https://maps.googleapis.com/maps/api/js?v=3.exp&libraries=geometry,drawing&key=AIzaSyCgbVaENPZQ1wOhdIyCok5yJTSVKBa6gYQ',
    loadingElement: (
      <div style={{ height: '400px' }}>
        <BarLoader />
      </div>
    ),
    containerElement: <div style={{ height: '400px' }} />,
    mapElement: <div style={{ height: '100%' }} />,
  }),
  withStateHandlers(
    () => ({
      isOpen: false,
    }),
    {
      onToggleOpen: ({ isOpen }) => () => ({
        isOpen: !isOpen,
      }),
    },
  ),
  lifecycle({
    componentWillMount() {
      const refs = {};

      this.setState({
        onCircleMounted: (ref) => {
          refs.circle = ref;
        },
        onCircleCenterChanged: () => {
          if (refs.circle) {
            console.log(refs.circle.getBounds());
            this.props.onCenterChanged(refs.circle.getCenter());
          }
        },
      });
    },
  }),
  withScriptjs,
  withGoogleMap,
)((props) => {
  const {
    advancedSearch,
    defaultCenter,
    defaultZoom,
    isOpen,
    markerName,
    onCircleCenterChanged,
    onCircleMounted,
    onDragEnd,
    onToggleOpen,
    radius,
  } = props;
  console.log('props: ', props);
  return (
    <GoogleMap defaultZoom={defaultZoom} defaultCenter={defaultCenter}>
      <Marker position={defaultCenter} onClick={onToggleOpen}>
        {isOpen && (
          <InfoBox onCloseClick={onToggleOpen} options={{ closeBoxURL: '', enableEventPropagation: true }}>
            <div style={{ backgroundColor: colors.info, padding: '12px' }}>
              <div style={{ fontSize: '16px', color: '#ffffff' }}>{markerName}</div>
            </div>
          </InfoBox>
        )}
      </Marker>
      {advancedSearch && (
        <Circle
          ref={onCircleMounted}
          defaultCenter={defaultCenter}
          defaultDraggable
          defaultRadius={20000}
          onDragEnd={onDragEnd}
          radius={parseInt(radius, 10)}
          onCenterChanged={onCircleCenterChanged}
          options={{ fillColor: colors.green, strokeColor: colors.green }}
        />
      )}
    </GoogleMap>
  );
});

type Props = {
  advancedSearch?: boolean,
  coordinates: CoordinatesObject,
  data: { name: string },
  onDragEnd?: Function,
  radius?: string,
  zoom?: number,
};
type State = {
  coordinates: ?CoordinatesObject,
  isMarkerShown: boolean,
};

class Map extends PureComponent<Props, State> {
  timeOut = null;

  static defaultProps = {
    advancedSearch: false,
    onDragEnd: () => {},
    radius: 10000,
    zoom: 11,
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      coordinates: null,
      isMarkerShown: false,
    };
  }

  componentDidMount() {
    this.delayedShowMarker();
  }

  componentWillUnmount() {
    if (this.timeOut) clearTimeout(this.timeOut);
  }

  delayedShowMarker = () => {
    this.timeOut = setTimeout(() => {
      this.setState({ isMarkerShown: true });
    }, 3000);
  };

  handleMarkerClick = () => {
    this.setState({ isMarkerShown: false });
    this.delayedShowMarker();
  };

  handleOnCircleDragEnd = () => {
    const { onDragEnd } = this.props;
    const { coordinates } = this.state;
    if (onDragEnd && coordinates) onDragEnd(coordinates);
  };

  onCircleCenterChanged = (coords: { lat: Function, lng: Function }) => {
    this.setState({ coordinates: { lat: coords.lat(), lng: coords.lng() } });
  };

  render() {
    const {
      advancedSearch, coordinates, data, radius, zoom,
    } = this.props;
    const { isMarkerShown } = this.state;
    const props = {
      advancedSearch,
      defaultCenter: coordinates,
      defaultZoom: zoom,
      isMarkerShown,
      markerName: data.name,
      onCenterChanged: this.onCircleCenterChanged,
      onDragEnd: this.handleOnCircleDragEnd,
      onMarkerClick: this.handleMarkerClick,
      radius,
    };
    return (
      <React.Fragment>
        <MyMapComponent {...props} />
      </React.Fragment>
    );
  }
}

export default Map;
