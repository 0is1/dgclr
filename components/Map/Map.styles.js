import styled from 'styled-components';

const Styles = {
  MapWrapper: styled.div`
    position: relative;
  `,
};

export default Styles;
