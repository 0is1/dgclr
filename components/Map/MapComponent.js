import React, { PureComponent } from 'react';
import {
  GoogleMap, useLoadScript, MarkerClusterer, Marker, InfoBox, Circle,
} from '@react-google-maps/api';
import Link from 'next/link';
import { isArrayWithLength } from 'helpers/utils';
import { BarLoader } from 'components/Spinners';
import colors from 'components/colors';
import type { CoordinatesObject, CourseForMap } from 'lib/types';

type MarkerData = { isOpen: boolean } & CourseForMap;

type Props = {
  advancedSearch: boolean,
  center: CoordinatesObject,
  data: { name: string, queryResults?: Array<?CourseForMap> },
  handleCenterChanged: Function,
  handleDragEnd: Function,
  handleMarkerClick: Function,
  handleMapCenterChange: Function,
  handleZoomChange: Function,
  markerName: string,
  markers: Array<?MarkerData>,
  radius: number,
  useCurrentLocation: boolean,
  zoom: number,
};

class Map extends PureComponent<Props> {
  constructor(props) {
    super(props);
    this.mapRef = React.createRef();
    this.circleRef = React.createRef();
  }

  state = {
    isOpen: false,
  };

  onToggleOpen = () => {
    const { isOpen } = this.state;
    this.setState({
      isOpen: !isOpen,
    });
  };

  onZoomChanged = () => {
    if (this.mapRef.current) {
      const { handleZoomChange } = this.props;
      const { state } = this.mapRef.current;
      // console.log('state.map.zoom: ', state.map.zoom);
      handleZoomChange(state.map.zoom);
    }
  };

  onMapCenterChanged = () => {
    if (this.mapRef.current) {
      const { handleMapCenterChange } = this.props;
      const { state } = this.mapRef.current;
      handleMapCenterChange(state.map.center);
    }
  };

  onCircleCenterChanged = () => {
    if (this.circleRef.current) {
      const { state } = this.circleRef.current;
      const { handleCenterChanged } = this.props;
      // console.log('state.circle.center: ', state.circle.center);
      handleCenterChanged(state.circle.center);
    }
  };

  onMarkerClustererClick = () => () => {
    // const clickedMarkers = markerClusterer.getMarkers();
    // console.log(`Current clicked markers length: ${clickedMarkers.length}`);
    // console.log(clickedMarkers);
  };

  render() {
    const {
      advancedSearch, center, zoom, markerName, markers, handleDragEnd, handleMarkerClick, radius, useCurrentLocation,
    } = this.props;
    const { isOpen } = this.state;
    const mapOptions = useCurrentLocation ? { center } : {};
    return (
      <div
        style={{
          height: '500px',
        }}
      >
        <GoogleMap
          id="example-map"
          zoom={zoom}
          center={center}
          onCenterChanged={this.onMapCenterChanged}
          onZoomChanged={this.onZoomChanged}
          mapContainerStyle={{
            height: '100%',
          }}
          ref={this.mapRef}
          options={mapOptions}
        >
          {!advancedSearch && (
            <Marker position={center} onClick={this.onToggleOpen}>
              {isOpen && (
                <InfoBox
                  onCloseClick={this.onToggleOpen}
                  options={{
                    closeBoxURL: '',
                    enableEventPropagation: true,
                  }}
                  position={center}
                >
                  <div
                    style={{
                      backgroundColor: colors.info,
                      padding: '12px',
                    }}
                  >
                    <div
                      style={{
                        fontSize: '16px',
                        color: '#ffffff',
                      }}
                    >
                      {markerName}
                    </div>
                  </div>
                </InfoBox>
              )}
            </Marker>
          )}
          {isArrayWithLength(markers) && (
            <MarkerClusterer onClick={this.onMarkerClustererClick} averageCenter enableRetinaIcons gridSize={60}>
              {clusterer => markers.map(marker => (
                <Marker
                  key={marker.name}
                  position={marker.coordinates}
                  onClick={() => {
                    handleMarkerClick(marker.id);
                  }}
                  clusterer={clusterer}
                >
                  {marker.isOpen && (
                  <InfoBox
                    onCloseClick={() => {
                      handleMarkerClick(marker.id);
                    }}
                    options={{
                      closeBoxURL: '',
                      enableEventPropagation: true,
                    }}
                    position={marker.coordinates}
                  >
                    <div
                      style={{
                        backgroundColor: colors.info,
                        padding: '10px',
                      }}
                    >
                      <div style={{ fontSize: '13px', color: '#ffffff' }}>{`${marker.name} – ${marker.address}`}</div>
                      <Link as={`/${marker.slug}`} href={`/course?slug=${marker.slug}`}>
                        <p>Näytä rata</p>
                      </Link>
                    </div>
                  </InfoBox>
                  )}
                </Marker>
              ))
              }
            </MarkerClusterer>
          )}
          {advancedSearch && (
            <Circle
              ref={this.circleRef}
              center={center}
              onDragEnd={handleDragEnd}
              onCenterChanged={this.onCircleCenterChanged}
              options={{
                fillColor: colors.green,
                strokeColor: colors.green,
                radius: parseInt(radius, 10),
                zIndex: 2,
                visible: true,
                draggable: true,
              }}
            />
          )}
        </GoogleMap>
      </div>
    );
  }
}

const LoadMap = (props) => {
  const { isLoaded, loadError } = useLoadScript({
    id: 'script-loader',
    googleMapsApiKey: 'AIzaSyCgbVaENPZQ1wOhdIyCok5yJTSVKBa6gYQ',
  });
  if (loadError) {
    return <div>Map cannot be loaded right now, sorry.</div>;
  }
  if (!isLoaded) {
    return (
      <div style={{ display: 'flex', justifyContent: 'center', height: '500px' }}>
        <BarLoader />
      </div>
    );
  }
  return <Map {...props} />;
};
export default LoadMap;
