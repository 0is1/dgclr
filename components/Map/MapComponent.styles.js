import styled from 'styled-components';
import colors from 'components/colors';

const Styles = {
  Wrapper: styled.div`
    height: 500px;
  `,
  LoaderWrapper: styled.div`
    display: flex;
    justify-content: center;
    height: 500px;
  `,
  InfoBoxContainer: styled.div`
    background-color: ${colors.info};
    padding: 10px;
  `,
  InfoBoxText: styled.p`
    fontsize: 13px;
    color: ${colors.white};
  `,
};

export default Styles;
