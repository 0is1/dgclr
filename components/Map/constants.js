export const MAP_RADIUS_DISTANCE_MAX = '150000';
export const MAP_RADIUS_DISTANCE_MIN = '1000';
