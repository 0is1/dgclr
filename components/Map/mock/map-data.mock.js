// eslint-disable-next-line import/prefer-default-export
export const filterQueryResultsForMapMock = [
  {
    address: 'Kilpailukuja, 04300 Tuusula',
    coordinates: { lat: 60.3905, lng: 25.03 },

    id: '5cfd0e367a3430892fcde89d',
    name: 'Ford DiscGolfPark',
    slug: 'ford-discgolfpark-tuusula',
  },
  {
    address: 'Keinukalliontie 47, 04250 Kerava',
    coordinates: {
      lat: 60.3846,
      lng: 25.1396,
    },
    id: '5cfd0e7f7a3430892fcde9c6',
    name: 'Keinukallio',
    slug: 'keinukallio-kerava',
  },
];
