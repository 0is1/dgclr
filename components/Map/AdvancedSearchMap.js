// @flow

import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Box, Label } from 'rebass';
import { debounce } from 'lodash';
import { setFilter as setFilterFunc, setAdvancedSearchMapZoom } from 'components/SearchContainer/actions';
import {
  getFilterTypeData,
  getAdvancedMapZoom,
  queryResultsFromState,
  latestAdvancedQuery as latestAdvancedQueryFunc,
} from 'components/SearchContainer/selectors';
import Map from 'components/Map';
import Input from 'components/Input';
import { ClipLoader } from 'components/Spinners';
import { ADVANCED_NEARBY } from 'lib/constants';
import { convertCoordinatesToObject, convertMetersToKilometers, courseAddressDetails } from 'helpers/utils';
import AdvancedSearchQueryStyles from 'components/SearchContainer/AdvancedSearchQuery.styles';

import type {
  Course, CoordinatesObject, CourseForMap, State as ReduxState,
} from 'lib/types';

const { NoResults } = AdvancedSearchQueryStyles;

type Props = {
  defaultValue: Array<{ coordinates: CoordinatesObject, radius: number }>,
  handleChange: Function,
  mapVisible: boolean,
  setFilter: Function,
  setMapZoom: Function,
  queryResults: Array<?Course>,
  zoom: number,
};
type State = {
  coordinates: CoordinatesObject,
  radius: string,
  waitingLocation: boolean,
  error: ?string,
};

const defaultCoordinates = { lat: 60.190599999999996, lng: 24.89741416931156 };

class AdvancedSearchMap extends Component<Props, State> {
  state = {
    coordinates: defaultCoordinates,
    radius: '20000',
    waitingLocation: true,
    error: null,
  };

  debounceRadius = debounce((radius) => {
    this.updateFilter({ radius });
  }, 300);

  inputOptions = {
    max: '150000',
    min: '1000',
    name: 'radius',
    step: '1000',
    type: 'range',
  };

  constructor(props) {
    super(props);
    const { defaultValue } = props;
    if (defaultValue.length && defaultValue[0].coordinates && defaultValue[0].radius) {
      this.state = {
        coordinates: defaultValue[0].coordinates,
        radius: defaultValue[0].radius,
        waitingLocation: true,
        error: null,
      };
    }
  }

  componentDidMount() {
    const { coordinates } = this.state;
    if (defaultCoordinates.lat === coordinates.lat && defaultCoordinates.lng === coordinates.lng) {
      try {
        const options = {
          enableHighAccuracy: true,
          timeout: 5000,
          maximumAge: 0,
        };

        const getCurrentPositionSuccess = (pos) => {
          const { coords } = pos;
          const newCoordinates = { lat: coords.latitude, lng: coords.longitude };
          this.updateFilter({ coordinates: newCoordinates });
          this.setState({ coordinates: newCoordinates, waitingLocation: false });
        };

        const getCurrentPositionError = (err) => {
          console.warn(`ERROR(${err.code}): ${err.message}`);
          this.setState({ waitingLocation: false, error: err.message });
        };
        navigator.geolocation.getCurrentPosition(getCurrentPositionSuccess, getCurrentPositionError, options);
      } catch (e) {
        console.error('getCurrentPosition error: ', e);

        this.setState({ waitingLocation: false });
      }
    } else {
      this.setState({ waitingLocation: false });
    }
  }

  componentDidUpdate(prevProps: Props) {
    const { mapVisible } = this.props;
    if (mapVisible && !prevProps.mapVisible) {
      this.updateFilter();
    }
  }

  updateFilter = (data: ?{} = {}) => {
    const { defaultValue, handleChange, setFilter } = this.props;
    const { coordinates, radius } = this.state;
    const [currentData = { coordinates, radius }] = defaultValue;
    const updatedFilter = { ...currentData, ...data };
    setFilter(ADVANCED_NEARBY, [updatedFilter]);
    handleChange(updatedFilter);
  };

  onCircleDragEnd = (coordinates: CoordinatesObject) => {
    this.updateFilter({ coordinates });
    this.setState({ coordinates });
  };

  onRadiusChange = (radius: string) => {
    this.setState({ radius });
    this.debounceRadius(radius);
  };

  filterQueryResultsForMap = (): Array<CourseForMap> => {
    const { queryResults } = this.props;
    return queryResults.map((result: ?Course) => {
      const {
        _id, name, locationInfo, slug,
      } = result || {};
      return {
        name,
        id: _id,
        address: courseAddressDetails(locationInfo),
        coordinates: convertCoordinatesToObject(locationInfo.location.coordinates),
        slug,
      };
    });
  };

  handleZoomChange = (zoom: number) => {
    const { setMapZoom } = this.props;
    if (parseInt(zoom, 10) > 0) {
      setMapZoom(zoom);
    }
  };

  render() {
    const { mapVisible, zoom } = this.props;
    if (!mapVisible) return null;
    const {
      coordinates, error, radius, waitingLocation,
    } = this.state;
    if (waitingLocation) {
      return <ClipLoader />;
    }
    const filteredResults = this.filterQueryResultsForMap();
    const props = {
      advancedSearch: true,
      coordinates,
      data: { name: `Keskipiste: ${coordinates.lat}, ${coordinates.lng}`, queryResults: filteredResults },
      onDragEnd: this.onCircleDragEnd,
      onZoomChange: this.handleZoomChange,
      radius,
      zoom,
    };

    return (
      <Fragment>
        {error && (
          <Box p={[0, '0.5rem 2rem']}>
            <NoResults>{error}</NoResults>
          </Box>
        )}
        <Map {...props} />
        <Label pt={['.5rem', '.5rem', '1rem', '1rem']}>{`Maksimiet√§isyys (${convertMetersToKilometers(parseInt(radius, 10))}km): `}</Label>
        <Input value={radius} options={this.inputOptions} onChange={this.onRadiusChange} />
      </Fragment>
    );
  }
}

const mapStateToProps = (state: ReduxState) => {
  const latestQuery = latestAdvancedQueryFunc(state);
  return {
    defaultValue: getFilterTypeData(state, ADVANCED_NEARBY),
    queryResults: queryResultsFromState(state, latestQuery),
    zoom: getAdvancedMapZoom(state),
  };
};
const mapDispatchToProps = dispatch => ({
  setFilter: (filterName, data) => dispatch(setFilterFunc(filterName, data)),
  setMapZoom: (zoom: number) => dispatch(setAdvancedSearchMapZoom(zoom)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearchMap);
