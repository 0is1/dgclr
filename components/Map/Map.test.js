import React from 'react';
import { shallow } from 'enzyme';
import Map from './Map';
import { defaultCoordinates as coordinates } from './AdvancedSearchMap';
import { filterQueryResultsForMapMock } from './mock/map-data.mock';

const mockProps = (props = {}) => ({
  advancedSearch: true,
  coordinates,
  data: {
    name: `Keskipiste: ${coordinates.lat}, ${coordinates.lng}`,
    queryResults: filterQueryResultsForMapMock,
  },
  onDragEnd: jest.fn(),
  onZoomChange: jest.fn(),
  radius: 20000,
  zoom: 9,
  ...props,
});

describe('<Map />', () => {
  describe('snapshots', () => {
    it('renders with default props', () => {
      const props = mockProps();
      const subject = shallow(<Map {...props} />);
      expect(subject).toMatchSnapshot();
    });
  });
});
