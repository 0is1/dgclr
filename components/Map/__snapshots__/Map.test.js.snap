// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> snapshots renders with default props 1`] = `
<LoadMap
  advancedSearch={true}
  center={
    Object {
      "lat": 60.190599999999996,
      "lng": 24.89741416931156,
    }
  }
  handleCenterChanged={[Function]}
  handleDragEnd={[Function]}
  handleOnMarkerClick={[Function]}
  handleZoomChange={[Function]}
  isMarkerShown={false}
  markerName="Keskipiste: 60.190599999999996, 24.89741416931156"
  markers={
    Array [
      Object {
        "address": "Kilpailukuja, 04300 Tuusula",
        "coordinates": Object {
          "lat": 60.3905,
          "lng": 25.03,
        },
        "id": "5cfd0e367a3430892fcde89d",
        "isOpen": false,
        "name": "Ford DiscGolfPark",
        "slug": "ford-discgolfpark-tuusula",
      },
      Object {
        "address": "Keinukalliontie 47, 04250 Kerava",
        "coordinates": Object {
          "lat": 60.3846,
          "lng": 25.1396,
        },
        "id": "5cfd0e7f7a3430892fcde9c6",
        "isOpen": false,
        "name": "Keinukallio",
        "slug": "keinukallio-kerava",
      },
    ]
  }
  radius={20000}
  zoom={9}
/>
`;
